// @effect/io@0.37.1/mjs/internal/metric/polling.mjs
import { dual } from "@effect/data/Function";
import { pipeArguments } from "@effect/data/Pipeable";
import * as core from "@effect/io/internal/core";
import * as circular from "@effect/io/internal/effect/circular";
import * as metric from "@effect/io/internal/metric";
import * as schedule from "@effect/io/internal/schedule";
/** @internal */
const PollingMetricSymbolKey = "@effect/io/Metric/Polling";
/** @internal */
export const PollingMetricTypeId = /*#__PURE__*/Symbol.for(PollingMetricSymbolKey);
/** @internal */
export const make = (metric, poll) => {
  return {
    [PollingMetricTypeId]: PollingMetricTypeId,
    pipe() {
      return pipeArguments(this, arguments);
    },
    metric,
    poll
  };
};
/** @internal */
export const collectAll = iterable => {
  const metrics = Array.from(iterable);
  return {
    [PollingMetricTypeId]: PollingMetricTypeId,
    pipe() {
      return pipeArguments(this, arguments);
    },
    metric: metric.make(Array.of(void 0), (inputs, extraTags) => {
      for (let i = 0; i < inputs.length; i++) {
        const pollingMetric = metrics[i];
        const input = (x => x[i])(inputs);
        pollingMetric.metric.unsafeUpdate(input, extraTags);
      }
    }, extraTags => Array.from(metrics.map(pollingMetric => pollingMetric.metric.unsafeValue(extraTags)))),
    poll: core.forEachSequential(metrics, metric => metric.poll)
  };
};
/** @internal */
export const launch = /*#__PURE__*/dual(2, (self, schedule) => circular.scheduleForked(schedule)(core.zipRight(metric.value(self.metric))(pollAndUpdate(self))));
/** @internal */
export const poll = self => self.poll;
/** @internal */
export const pollAndUpdate = self => core.flatMap(self.poll, value => metric.update(self.metric, value));
/** @internal */
export const retry = /*#__PURE__*/dual(2, (self, policy) => ({
  [PollingMetricTypeId]: PollingMetricTypeId,
  pipe() {
    return pipeArguments(this, arguments);
  },
  metric: self.metric,
  poll: schedule.retry_Effect(self.poll, policy)
}));
/** @internal */
export const zip = /*#__PURE__*/dual(2, (self, that) => ({
  [PollingMetricTypeId]: PollingMetricTypeId,
  pipe() {
    return pipeArguments(this, arguments);
  },
  metric: metric.zip(that.metric)(self.metric),
  poll: core.zip(self.poll, that.poll)
}));
