// @effect/io@0.37.1/Clock.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sleep = exports.make = exports.currentTimeNanos = exports.currentTimeMillis = exports.clockWith = exports.ClockTypeId = exports.Clock = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/clock"));
var defaultServices = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/defaultServices"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const ClockTypeId = internal.ClockTypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.ClockTypeId = ClockTypeId;
const make = internal.make;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.make = make;
const sleep = defaultServices.sleep;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.sleep = sleep;
const currentTimeMillis = defaultServices.currentTimeMillis;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.currentTimeMillis = currentTimeMillis;
const currentTimeNanos = defaultServices.currentTimeNanos;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.currentTimeNanos = currentTimeNanos;
const clockWith = defaultServices.clockWith;
/**
 * @since 1.0.0
 * @category context
 */
exports.clockWith = clockWith;
const Clock = internal.clockTag;
exports.Clock = Clock;
