// @effect/io@0.37.1/FiberRef.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateSomeAndGet = exports.updateSome = exports.updateAndGet = exports.update = exports.unsafeMakeSupervisor = exports.unsafeMakePatch = exports.unsafeMakeHashSet = exports.unsafeMakeContext = exports.unsafeMake = exports.unhandledErrorLogLevel = exports.set = exports.reset = exports.modifySome = exports.modify = exports.makeWith = exports.makeRuntimeFlags = exports.makeContext = exports.make = exports.interruptedCause = exports.getWith = exports.getAndUpdateSome = exports.getAndUpdate = exports.getAndSet = exports.get = exports.delete = exports.currentTracerTimingEnabled = exports.currentTracerSpanLinks = exports.currentTracerSpanAnnotations = exports.currentTracerSpan = exports.currentSupervisor = exports.currentSchedulingPriority = exports.currentScheduler = exports.currentRuntimeFlags = exports.currentRequestCacheEnabled = exports.currentRequestCache = exports.currentRequestBatchingEnabled = exports.currentMinimumLogLevel = exports.currentMetricLabels = exports.currentMaxFiberOps = exports.currentLoggers = exports.currentLogSpan = exports.currentLogLevel = exports.currentLogAnnotations = exports.currentContext = exports.FiberRefTypeId = void 0;
var core = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/core"));
var fiberRuntime = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/fiberRuntime"));
var query = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/query"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const FiberRefTypeId = core.FiberRefTypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.FiberRefTypeId = FiberRefTypeId;
const make = fiberRuntime.fiberRefMake;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.make = make;
const makeWith = fiberRuntime.fiberRefMakeWith;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.makeWith = makeWith;
const makeContext = fiberRuntime.fiberRefMakeContext;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.makeContext = makeContext;
const makeRuntimeFlags = fiberRuntime.fiberRefMakeRuntimeFlags;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.makeRuntimeFlags = makeRuntimeFlags;
const unsafeMake = core.fiberRefUnsafeMake;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.unsafeMake = unsafeMake;
const unsafeMakeHashSet = core.fiberRefUnsafeMakeHashSet;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.unsafeMakeHashSet = unsafeMakeHashSet;
const unsafeMakeContext = core.fiberRefUnsafeMakeContext;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.unsafeMakeContext = unsafeMakeContext;
const unsafeMakeSupervisor = fiberRuntime.fiberRefUnsafeMakeSupervisor;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.unsafeMakeSupervisor = unsafeMakeSupervisor;
const unsafeMakePatch = core.fiberRefUnsafeMakePatch;
/**
 * @since 1.0.0
 * @category getters
 */
exports.unsafeMakePatch = unsafeMakePatch;
const get = core.fiberRefGet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.get = get;
const getAndSet = core.fiberRefGetAndSet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndSet = getAndSet;
const getAndUpdate = core.fiberRefGetAndUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdate = getAndUpdate;
const getAndUpdateSome = core.fiberRefGetAndUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateSome = getAndUpdateSome;
const getWith = core.fiberRefGetWith;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getWith = getWith;
const set = core.fiberRefSet;
exports.set = set;
const _delete = core.fiberRefDelete;
exports.delete = _delete;
/**
 * @since 1.0.0
 * @category utils
 */
const reset = core.fiberRefReset;
/**
 * @since 1.0.0
 * @category utils
 */
exports.reset = reset;
const modify = core.fiberRefModify;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modify = modify;
const modifySome = core.fiberRefModifySome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modifySome = modifySome;
const update = core.fiberRefUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
exports.update = update;
const updateSome = core.fiberRefUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateSome = updateSome;
const updateAndGet = core.fiberRefUpdateAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateAndGet = updateAndGet;
const updateSomeAndGet = core.fiberRefUpdateSomeAndGet;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.updateSomeAndGet = updateSomeAndGet;
const currentRequestBatchingEnabled = core.currentRequestBatching;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentRequestBatchingEnabled = currentRequestBatchingEnabled;
const currentRequestCache = query.currentCache;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentRequestCache = currentRequestCache;
const currentRequestCacheEnabled = query.currentCacheEnabled;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentRequestCacheEnabled = currentRequestCacheEnabled;
const currentContext = core.currentContext;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentContext = currentContext;
const currentSchedulingPriority = core.currentSchedulingPriority;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentSchedulingPriority = currentSchedulingPriority;
const currentMaxFiberOps = core.currentMaxFiberOps;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentMaxFiberOps = currentMaxFiberOps;
const unhandledErrorLogLevel = core.currentUnhandledErrorLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.unhandledErrorLogLevel = unhandledErrorLogLevel;
const currentLogAnnotations = core.currentLogAnnotations;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentLogAnnotations = currentLogAnnotations;
const currentLoggers = fiberRuntime.currentLoggers;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentLoggers = currentLoggers;
const currentLogLevel = core.currentLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentLogLevel = currentLogLevel;
const currentMinimumLogLevel = fiberRuntime.currentMinimumLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentMinimumLogLevel = currentMinimumLogLevel;
const currentLogSpan = core.currentLogSpan;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentLogSpan = currentLogSpan;
const currentRuntimeFlags = fiberRuntime.currentRuntimeFlags;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentRuntimeFlags = currentRuntimeFlags;
const currentScheduler = core.currentScheduler;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentScheduler = currentScheduler;
const currentSupervisor = fiberRuntime.currentSupervisor;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentSupervisor = currentSupervisor;
const currentMetricLabels = core.currentMetricLabels;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentMetricLabels = currentMetricLabels;
const currentTracerSpan = core.currentTracerSpan;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentTracerSpan = currentTracerSpan;
const currentTracerTimingEnabled = core.currentTracerTimingEnabled;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentTracerTimingEnabled = currentTracerTimingEnabled;
const currentTracerSpanAnnotations = core.currentTracerSpanAnnotations;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentTracerSpanAnnotations = currentTracerSpanAnnotations;
const currentTracerSpanLinks = core.currentTracerSpanLinks;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
exports.currentTracerSpanLinks = currentTracerSpanLinks;
const interruptedCause = core.currentInterruptedCause;
exports.interruptedCause = interruptedCause;
