// @effect/io@0.37.1/Ref/Synchronized.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateSomeEffect = exports.updateSomeAndGetEffect = exports.updateSomeAndGet = exports.updateSome = exports.updateEffect = exports.updateAndGetEffect = exports.updateAndGet = exports.update = exports.unsafeMake = exports.setAndGet = exports.set = exports.modifySomeEffect = exports.modifySome = exports.modifyEffect = exports.modify = exports.make = exports.getAndUpdateSomeEffect = exports.getAndUpdateSome = exports.getAndUpdateEffect = exports.getAndUpdate = exports.getAndSet = exports.get = exports.SynchronizedTypeId = void 0;
var circular = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/effect/circular"));
var ref = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/ref"));
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("@effect/io/internal/synchronizedRef"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * @since 1.0.0
 * @category symbols
 */
const SynchronizedTypeId = circular.SynchronizedTypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
exports.SynchronizedTypeId = SynchronizedTypeId;
const make = circular.makeSynchronized;
/**
 * @since 1.0.0
 * @category getters
 */
exports.make = make;
const get = ref.get;
/**
 * @since 1.0.0
 * @category utils
 */
exports.get = get;
const getAndSet = ref.getAndSet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndSet = getAndSet;
const getAndUpdate = ref.getAndUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdate = getAndUpdate;
const getAndUpdateEffect = internal.getAndUpdateEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateEffect = getAndUpdateEffect;
const getAndUpdateSome = ref.getAndUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateSome = getAndUpdateSome;
const getAndUpdateSomeEffect = internal.getAndUpdateSomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.getAndUpdateSomeEffect = getAndUpdateSomeEffect;
const modify = internal.modify;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modify = modify;
const modifyEffect = internal.modifyEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modifyEffect = modifyEffect;
const modifySome = ref.modifySome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modifySome = modifySome;
const modifySomeEffect = internal.modifySomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.modifySomeEffect = modifySomeEffect;
const set = ref.set;
/**
 * @since 1.0.0
 * @category utils
 */
exports.set = set;
const setAndGet = ref.setAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.setAndGet = setAndGet;
const update = ref.update;
/**
 * @since 1.0.0
 * @category utils
 */
exports.update = update;
const updateEffect = internal.updateEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateEffect = updateEffect;
const updateAndGet = ref.updateAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateAndGet = updateAndGet;
const updateAndGetEffect = internal.updateAndGetEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateAndGetEffect = updateAndGetEffect;
const updateSome = ref.updateSome;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateSome = updateSome;
const updateSomeEffect = internal.updateSomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateSomeEffect = updateSomeEffect;
const updateSomeAndGet = ref.updateSomeAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
exports.updateSomeAndGet = updateSomeAndGet;
const updateSomeAndGetEffect = circular.updateSomeAndGetEffectSynchronized;
/**
 * @since 1.0.0
 * @category unsafe
 */
exports.updateSomeAndGetEffect = updateSomeAndGetEffect;
const unsafeMake = circular.unsafeMakeSynchronized;
exports.unsafeMake = unsafeMake;
