// @effect/io@0.37.1/mjs/internal/scopedRef.mjs
import * as Context from "@effect/data/Context";
import { dual } from "@effect/data/Function";
import { pipeArguments } from "@effect/data/Pipeable";
import * as core from "@effect/io/internal/core";
import * as effect from "@effect/io/internal/effect";
import * as circular from "@effect/io/internal/effect/circular";
import * as fiberRuntime from "@effect/io/internal/fiberRuntime";
import * as ref from "@effect/io/internal/ref";
import * as synchronized from "@effect/io/internal/synchronizedRef";
/** @internal */
const ScopedRefSymbolKey = "@effect/io/ScopedRef";
/** @internal */
export const ScopedRefTypeId = /*#__PURE__*/Symbol.for(ScopedRefSymbolKey);
/** @internal */
const scopedRefVariance = {
  _A: _ => _
};
/** @internal  */
const close = self => core.flatMap(ref.get(self.ref), tuple => tuple[0].close(core.exitUnit));
/** @internal */
export const fromAcquire = acquire => core.uninterruptibleMask(restore => core.flatMap(newScope => core.flatMap(value => core.flatMap(ref => {
  const scopedRef = {
    [ScopedRefTypeId]: scopedRefVariance,
    pipe() {
      return pipeArguments(this, arguments);
    },
    ref
  };
  return core.as(scopedRef)(fiberRuntime.addFinalizer(() => close(scopedRef)));
})(circular.makeSynchronized([newScope, value])))(core.onError(cause => newScope.close(core.exitFail(cause)))(restore(core.mapInputContext(Context.add(fiberRuntime.scopeTag, newScope))(acquire)))))(fiberRuntime.scopeMake()));
/** @internal */
export const get = self => core.map(ref.get(self.ref), tuple => tuple[1]);
/** @internal */
export const make = evaluate => fromAcquire(core.sync(evaluate));
/** @internal */
export const set = /*#__PURE__*/dual(2, (self, acquire) => core.flatten(synchronized.modifyEffect(self.ref, ([oldScope, value]) => core.uninterruptibleMask(restore => core.flatMap(newScope => core.flatMap(core.exitMatch({
  onFailure: cause => core.as([core.failCause(cause), [oldScope, value]])(effect.ignore(newScope.close(core.exitUnit))),
  onSuccess: value => core.as([core.unit, [newScope, value]])(effect.ignore(oldScope.close(core.exitUnit)))
}))(core.exit(restore(core.mapInputContext(Context.add(fiberRuntime.scopeTag, newScope))(acquire)))))(fiberRuntime.scopeMake())))));
