// @substreams/core@0.16.0/dist/cjs/proto/sf/substreams/v1/clock_pb.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Clock = exports.BlockRef = void 0;
var _protobuf = /*#__PURE__*/require("@bufbuild/protobuf");
// @generated by protoc-gen-es v1.7.2 with parameter "target=ts"
// @generated from file sf/substreams/v1/clock.proto (package sf.substreams.v1, syntax proto3)
/* eslint-disable */
// @ts-nocheck

/**
 * Clock is a pointer to a block with added timestamp
 *
 * @generated from message sf.substreams.v1.Clock
 */
class Clock extends _protobuf.Message {
  /**
   * @generated from field: string id = 1;
   */
  id = "";
  /**
   * @generated from field: uint64 number = 2;
   */
  number = _protobuf.protoInt64.zero;
  /**
   * @generated from field: google.protobuf.Timestamp timestamp = 3;
   */
  timestamp;
  constructor(data) {
    super();
    _protobuf.proto3.util.initPartial(data, this);
  }
  static runtime = _protobuf.proto3;
  static typeName = "sf.substreams.v1.Clock";
  static fields = _protobuf.proto3.util.newFieldList(() => [{
    no: 1,
    name: "id",
    kind: "scalar",
    T: 9 /* ScalarType.STRING */
  }, {
    no: 2,
    name: "number",
    kind: "scalar",
    T: 4 /* ScalarType.UINT64 */
  }, {
    no: 3,
    name: "timestamp",
    kind: "message",
    T: _protobuf.Timestamp
  }]);
  static fromBinary(bytes, options) {
    return new Clock().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new Clock().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new Clock().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return _protobuf.proto3.util.equals(Clock, a, b);
  }
}
/**
 * BlockRef is a pointer to a block to which we don't know the timestamp
 *
 * @generated from message sf.substreams.v1.BlockRef
 */
exports.Clock = Clock;
class BlockRef extends _protobuf.Message {
  /**
   * @generated from field: string id = 1;
   */
  id = "";
  /**
   * @generated from field: uint64 number = 2;
   */
  number = _protobuf.protoInt64.zero;
  constructor(data) {
    super();
    _protobuf.proto3.util.initPartial(data, this);
  }
  static runtime = _protobuf.proto3;
  static typeName = "sf.substreams.v1.BlockRef";
  static fields = _protobuf.proto3.util.newFieldList(() => [{
    no: 1,
    name: "id",
    kind: "scalar",
    T: 9 /* ScalarType.STRING */
  }, {
    no: 2,
    name: "number",
    kind: "scalar",
    T: 4 /* ScalarType.UINT64 */
  }]);
  static fromBinary(bytes, options) {
    return new BlockRef().fromBinary(bytes, options);
  }
  static fromJson(jsonValue, options) {
    return new BlockRef().fromJson(jsonValue, options);
  }
  static fromJsonString(jsonString, options) {
    return new BlockRef().fromJsonString(jsonString, options);
  }
  static equals(a, b) {
    return _protobuf.proto3.util.equals(BlockRef, a, b);
  }
}
exports.BlockRef = BlockRef;
