// @effect/io@0.37.1/mjs/FiberRef.mjs
import * as core from "@effect/io/internal/core";
import * as fiberRuntime from "@effect/io/internal/fiberRuntime";
import * as query from "@effect/io/internal/query";
/**
 * @since 1.0.0
 * @category symbols
 */
export const FiberRefTypeId = core.FiberRefTypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = fiberRuntime.fiberRefMake;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeWith = fiberRuntime.fiberRefMakeWith;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeContext = fiberRuntime.fiberRefMakeContext;
/**
 * @since 1.0.0
 * @category constructors
 */
export const makeRuntimeFlags = fiberRuntime.fiberRefMakeRuntimeFlags;
/**
 * @since 1.0.0
 * @category constructors
 */
export const unsafeMake = core.fiberRefUnsafeMake;
/**
 * @since 1.0.0
 * @category constructors
 */
export const unsafeMakeHashSet = core.fiberRefUnsafeMakeHashSet;
/**
 * @since 1.0.0
 * @category constructors
 */
export const unsafeMakeContext = core.fiberRefUnsafeMakeContext;
/**
 * @since 1.0.0
 * @category constructors
 */
export const unsafeMakeSupervisor = fiberRuntime.fiberRefUnsafeMakeSupervisor;
/**
 * @since 1.0.0
 * @category constructors
 */
export const unsafeMakePatch = core.fiberRefUnsafeMakePatch;
/**
 * @since 1.0.0
 * @category getters
 */
export const get = core.fiberRefGet;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndSet = core.fiberRefGetAndSet;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdate = core.fiberRefGetAndUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateSome = core.fiberRefGetAndUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
export const getWith = core.fiberRefGetWith;
/**
 * @since 1.0.0
 * @category utils
 */
export const set = core.fiberRefSet;
const _delete = core.fiberRefDelete;
export {
/**
 * @since 1.0.0
 * @category utils
 */
_delete as delete };
/**
 * @since 1.0.0
 * @category utils
 */
export const reset = core.fiberRefReset;
/**
 * @since 1.0.0
 * @category utils
 */
export const modify = core.fiberRefModify;
/**
 * @since 1.0.0
 * @category utils
 */
export const modifySome = core.fiberRefModifySome;
/**
 * @since 1.0.0
 * @category utils
 */
export const update = core.fiberRefUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSome = core.fiberRefUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateAndGet = core.fiberRefUpdateAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSomeAndGet = core.fiberRefUpdateSomeAndGet;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentRequestBatchingEnabled = core.currentRequestBatching;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentRequestCache = query.currentCache;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentRequestCacheEnabled = query.currentCacheEnabled;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentContext = core.currentContext;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentSchedulingPriority = core.currentSchedulingPriority;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentMaxFiberOps = core.currentMaxFiberOps;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const unhandledErrorLogLevel = core.currentUnhandledErrorLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentLogAnnotations = core.currentLogAnnotations;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentLoggers = fiberRuntime.currentLoggers;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentLogLevel = core.currentLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentMinimumLogLevel = fiberRuntime.currentMinimumLogLevel;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentLogSpan = core.currentLogSpan;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentRuntimeFlags = fiberRuntime.currentRuntimeFlags;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentScheduler = core.currentScheduler;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentSupervisor = fiberRuntime.currentSupervisor;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentMetricLabels = core.currentMetricLabels;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentTracerSpan = core.currentTracerSpan;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentTracerTimingEnabled = core.currentTracerTimingEnabled;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentTracerSpanAnnotations = core.currentTracerSpanAnnotations;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const currentTracerSpanLinks = core.currentTracerSpanLinks;
/**
 * @since 1.0.0
 * @category fiberRefs
 */
export const interruptedCause = core.currentInterruptedCause;
