// @effect/io@0.37.1/mjs/internal/testing/annotations.mjs
var _a;
import * as Context from "@effect/data/Context";
import * as Equal from "@effect/data/Equal";
import * as MutableRef from "@effect/data/MutableRef";
import * as RA from "@effect/data/ReadonlyArray";
import * as SortedSet from "@effect/data/SortedSet";
import * as core from "@effect/io/internal/core";
import * as effect from "@effect/io/internal/effect";
import * as fiber from "@effect/io/internal/fiber";
import * as TestAnnotation from "@effect/io/internal/testing/testAnnotation";
import * as TestAnnotationMap from "@effect/io/internal/testing/testAnnotationMap";
import * as Ref from "@effect/io/Ref";
/** @internal */
export const AnnotationsTypeId = /*#__PURE__*/Symbol.for("@effect/test/Annotations");
/** @internal */
class AnnotationsImpl {
  constructor(ref) {
    this.ref = ref;
    this[_a] = AnnotationsTypeId;
  }
  get(key) {
    return core.map(Ref.get(this.ref), TestAnnotationMap.get(key));
  }
  annotate(key, value) {
    return Ref.update(this.ref, TestAnnotationMap.annotate(key, value));
  }
  supervisedFibers() {
    return effect.descriptorWith(descriptor => core.flatMap(this.get(TestAnnotation.fibers), either => {
      switch (either._tag) {
        case "Left":
          {
            return core.succeed(SortedSet.empty(fiber.Order));
          }
        case "Right":
          {
            return core.map(SortedSet.filter(fiber => !Equal.equals(fiber.id(), descriptor.id)))(core.map(RA.reduce(SortedSet.empty(fiber.Order), (a, b) => SortedSet.union(a, b)))(core.forEachSequential(ref => core.sync(() => MutableRef.get(ref)))(either.right)));
          }
      }
    }));
  }
}
_a = AnnotationsTypeId;
/** @internal */
export const Annotations = /*#__PURE__*/Context.Tag( /*#__PURE__*/Symbol.for("@effect/test/Annotations"));
/** @internal */
export const isAnnotations = u => {
  return typeof u === "object" && u != null && AnnotationsTypeId in u;
};
/** @internal */
export const make = ref => new AnnotationsImpl(ref);
