// @substreams/core@0.16.0/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ModuleGraph", {
  enumerable: true,
  get: function () {
    return _createModuleGraph.ModuleGraph;
  }
});
Object.defineProperty(exports, "applyParams", {
  enumerable: true,
  get: function () {
    return _applyParams.applyParams;
  }
});
Object.defineProperty(exports, "authIssue", {
  enumerable: true,
  get: function () {
    return _authIssue.authIssue;
  }
});
Object.defineProperty(exports, "calculateHeadBlockTimeDrift", {
  enumerable: true,
  get: function () {
    return _calculateHeadBlockTimeDrift.calculateHeadBlockTimeDrift;
  }
});
Object.defineProperty(exports, "createAuthInterceptor", {
  enumerable: true,
  get: function () {
    return _createAuthInterceptor.createAuthInterceptor;
  }
});
Object.defineProperty(exports, "createHash", {
  enumerable: true,
  get: function () {
    return _createHash.createHash;
  }
});
Object.defineProperty(exports, "createModuleGraph", {
  enumerable: true,
  get: function () {
    return _createModuleGraph.createModuleGraph;
  }
});
Object.defineProperty(exports, "createModuleHash", {
  enumerable: true,
  get: function () {
    return _createModuleHash.createModuleHash;
  }
});
Object.defineProperty(exports, "createModuleHashHex", {
  enumerable: true,
  get: function () {
    return _createModuleHash.createModuleHashHex;
  }
});
Object.defineProperty(exports, "createRegistry", {
  enumerable: true,
  get: function () {
    return _createRegistry.createRegistry;
  }
});
Object.defineProperty(exports, "createRequest", {
  enumerable: true,
  get: function () {
    return _createRequest.createRequest;
  }
});
Object.defineProperty(exports, "createSubstream", {
  enumerable: true,
  get: function () {
    return _createSubstream.createSubstream;
  }
});
Object.defineProperty(exports, "fetchSubstream", {
  enumerable: true,
  get: function () {
    return _fetchSubstream.fetchSubstream;
  }
});
Object.defineProperty(exports, "generateMermaidGraph", {
  enumerable: true,
  get: function () {
    return _generateMermaidGraph.generateMermaidGraph;
  }
});
Object.defineProperty(exports, "getModule", {
  enumerable: true,
  get: function () {
    return _getModule.getModule;
  }
});
Object.defineProperty(exports, "getModuleOrThrow", {
  enumerable: true,
  get: function () {
    return _getModule.getModuleOrThrow;
  }
});
Object.defineProperty(exports, "getModules", {
  enumerable: true,
  get: function () {
    return _getModules.getModules;
  }
});
Object.defineProperty(exports, "getOutputType", {
  enumerable: true,
  get: function () {
    return _getOutputType.getOutputType;
  }
});
Object.defineProperty(exports, "getProtoType", {
  enumerable: true,
  get: function () {
    return _getProtoType.getProtoType;
  }
});
Object.defineProperty(exports, "getProtoTypeName", {
  enumerable: true,
  get: function () {
    return _getProtoTypeName.getProtoTypeName;
  }
});
Object.defineProperty(exports, "getProtoTypeOrThrow", {
  enumerable: true,
  get: function () {
    return _getProtoType.getProtoTypeOrThrow;
  }
});
Object.defineProperty(exports, "isEmptyMessage", {
  enumerable: true,
  get: function () {
    return _isEmptyMessage.isEmptyMessage;
  }
});
Object.defineProperty(exports, "isMapModule", {
  enumerable: true,
  get: function () {
    return _isMapModule.isMapModule;
  }
});
Object.defineProperty(exports, "isStoreModule", {
  enumerable: true,
  get: function () {
    return _isStoreModule.isStoreModule;
  }
});
Object.defineProperty(exports, "mergeProgressRanges", {
  enumerable: true,
  get: function () {
    return _mergeProgressRanges.mergeProgressRanges;
  }
});
Object.defineProperty(exports, "mergeSortedProgressRanges", {
  enumerable: true,
  get: function () {
    return _mergeProgressRanges.mergeSortedProgressRanges;
  }
});
Object.defineProperty(exports, "nameRegExp", {
  enumerable: true,
  get: function () {
    return _validatePackage.nameRegExp;
  }
});
Object.defineProperty(exports, "parseAuthorization", {
  enumerable: true,
  get: function () {
    return _authIssue.parseAuthorization;
  }
});
Object.defineProperty(exports, "semverRegExp", {
  enumerable: true,
  get: function () {
    return _validatePackage.semverRegExp;
  }
});
Object.defineProperty(exports, "storeModeName", {
  enumerable: true,
  get: function () {
    return _storeModeName.storeModeName;
  }
});
Object.defineProperty(exports, "streamBlocks", {
  enumerable: true,
  get: function () {
    return _streamBlocks.streamBlocks;
  }
});
Object.defineProperty(exports, "toHex", {
  enumerable: true,
  get: function () {
    return _toHex.toHex;
  }
});
Object.defineProperty(exports, "unpackMapOutput", {
  enumerable: true,
  get: function () {
    return _unpackMapOutput.unpackMapOutput;
  }
});
Object.defineProperty(exports, "validateModules", {
  enumerable: true,
  get: function () {
    return _validateModules.validateModules;
  }
});
Object.defineProperty(exports, "validatePackage", {
  enumerable: true,
  get: function () {
    return _validatePackage.validatePackage;
  }
});
Object.defineProperty(exports, "validatePackageAndModules", {
  enumerable: true,
  get: function () {
    return _validatePackageAndModules.validatePackageAndModules;
  }
});
var _createAuthInterceptor = /*#__PURE__*/require("./auth/create-auth-interceptor.js");
var _createRegistry = /*#__PURE__*/require("./utils/create-registry.js");
var _createRequest = /*#__PURE__*/require("./utils/create-request.js");
var _createSubstream = /*#__PURE__*/require("./utils/create-substream.js");
var _getModule = /*#__PURE__*/require("./utils/get-module.js");
var _getModules = /*#__PURE__*/require("./utils/get-modules.js");
var _getOutputType = /*#__PURE__*/require("./utils/get-output-type.js");
var _getProtoType = /*#__PURE__*/require("./utils/get-proto-type.js");
var _getProtoTypeName = /*#__PURE__*/require("./utils/get-proto-type-name.js");
var _isMapModule = /*#__PURE__*/require("./utils/is-map-module.js");
var _isStoreModule = /*#__PURE__*/require("./utils/is-store-module.js");
var _streamBlocks = /*#__PURE__*/require("./utils/stream-blocks.js");
var _unpackMapOutput = /*#__PURE__*/require("./utils/unpack-map-output.js");
var _isEmptyMessage = /*#__PURE__*/require("./utils/is-empty-message.js");
var _mergeProgressRanges = /*#__PURE__*/require("./utils/merge-progress-ranges.js");
var _fetchSubstream = /*#__PURE__*/require("./utils/fetch-substream.js");
var _storeModeName = /*#__PURE__*/require("./utils/store-mode-name.js");
var _toHex = /*#__PURE__*/require("./utils/to-hex.js");
var _createHash = /*#__PURE__*/require("./utils/create-hash.js");
var _authIssue = /*#__PURE__*/require("./auth/auth-issue.js");
var _createModuleGraph = /*#__PURE__*/require("./manifest/graph/create-module-graph.js");
var _createModuleHash = /*#__PURE__*/require("./manifest/signature/create-module-hash.js");
var _applyParams = /*#__PURE__*/require("./manifest/params/apply-params.js");
var _validatePackage = /*#__PURE__*/require("./manifest/validation/validate-package.js");
var _validateModules = /*#__PURE__*/require("./manifest/validation/validate-modules.js");
var _validatePackageAndModules = /*#__PURE__*/require("./manifest/validation/validate-package-and-modules.js");
var _calculateHeadBlockTimeDrift = /*#__PURE__*/require("./utils/calculate-head-block-time-drift.js");
var _generateMermaidGraph = /*#__PURE__*/require("./utils/generate-mermaid-graph.js");
