// @effect/io@0.37.1/mjs/Ref/Synchronized.mjs
import * as circular from "@effect/io/internal/effect/circular";
import * as ref from "@effect/io/internal/ref";
import * as internal from "@effect/io/internal/synchronizedRef";
/**
 * @since 1.0.0
 * @category symbols
 */
export const SynchronizedTypeId = circular.SynchronizedTypeId;
/**
 * @since 1.0.0
 * @category constructors
 */
export const make = circular.makeSynchronized;
/**
 * @since 1.0.0
 * @category getters
 */
export const get = ref.get;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndSet = ref.getAndSet;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdate = ref.getAndUpdate;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateEffect = internal.getAndUpdateEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateSome = ref.getAndUpdateSome;
/**
 * @since 1.0.0
 * @category utils
 */
export const getAndUpdateSomeEffect = internal.getAndUpdateSomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const modify = internal.modify;
/**
 * @since 1.0.0
 * @category utils
 */
export const modifyEffect = internal.modifyEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const modifySome = ref.modifySome;
/**
 * @since 1.0.0
 * @category utils
 */
export const modifySomeEffect = internal.modifySomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const set = ref.set;
/**
 * @since 1.0.0
 * @category utils
 */
export const setAndGet = ref.setAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
export const update = ref.update;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateEffect = internal.updateEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateAndGet = ref.updateAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateAndGetEffect = internal.updateAndGetEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSome = ref.updateSome;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSomeEffect = internal.updateSomeEffect;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSomeAndGet = ref.updateSomeAndGet;
/**
 * @since 1.0.0
 * @category utils
 */
export const updateSomeAndGetEffect = circular.updateSomeAndGetEffectSynchronized;
/**
 * @since 1.0.0
 * @category unsafe
 */
export const unsafeMake = circular.unsafeMakeSynchronized;
